src/rdkafka_telemetry_encode.c

```diff
@@ -0,0 +1,834 @@
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rd.h"
+#include "rdkafka_int.h"
+#include "rdkafka_telemetry_encode.h"
+#include "nanopb/pb.h"
+#include "nanopb/pb_encode.h"
+#include "nanopb/pb_decode.h"
+#include "opentelemetry/metrics.pb.h"
+
+#define THREE_ORDERS_MAGNITUDE 1000
+
+typedef struct {
+        opentelemetry_proto_metrics_v1_Metric **metrics;
+        size_t count;
+} rd_kafka_telemetry_metrics_repeated_t;
+
+typedef struct {
+        opentelemetry_proto_common_v1_KeyValue **key_values;
+        size_t count;
+} rd_kafka_telemetry_key_values_repeated_t;
+
+
+static rd_kafka_telemetry_metric_value_t
+calculate_connection_creation_total(rd_kafka_t *rk,
+                                    rd_kafka_broker_t *rkb_selected,
+                                    rd_ts_t now_ns) {
+        rd_kafka_telemetry_metric_value_t total;
+        rd_kafka_broker_t *rkb;
+
+        total.int_value = 0;
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                const int32_t connects = rd_atomic32_get(&rkb->rkb_c.connects);
+                if (!rk->rk_telemetry.delta_temporality)
+                        total.int_value += connects;
+                else
+                        total.int_value +=
+                            connects -
+                            rkb->rkb_telemetry.rkb_historic_c.connects;
+        }
+
+        return total;
+}
+
+static rd_kafka_telemetry_metric_value_t
+calculate_connection_creation_rate(rd_kafka_t *rk,
+                                   rd_kafka_broker_t *rkb_selected,
+                                   rd_ts_t now_ns) {
+        rd_kafka_telemetry_metric_value_t total;
+        rd_kafka_broker_t *rkb;
+        rd_ts_t ts_last = rk->rk_telemetry.rk_historic_c.ts_last;
+
+        total.double_value = 0;
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                total.double_value +=
+                    rd_atomic32_get(&rkb->rkb_c.connects) -
+                    rkb->rkb_telemetry.rkb_historic_c.connects;
+        }
+        double seconds = (now_ns - ts_last) / 1e9;
+        if (seconds > 0)
```

Comment:
In case the client is shutting down, interval is less that one second and the rate increases from let's say 3 connections/s to 52 connections/s because of the denominator that is less than one. I'd avoid this as it's an outlier by dividing only if greater than 1
```suggestion
        if (seconds > 1.0)
```

Reply:
Makes sense, updated